daemon/logger/local/read.go

```diff
@@ -99,63 +98,71 @@ func getTailReader(ctx context.Context, r loggerutils.SizeReaderAt, req int) (io
 type decoder struct {
 	rdr   io.Reader
 	proto *logdriver.LogEntry
-	buf   []byte
+	// buf keeps bytes from rdr.
+	buf []byte
+	// offset is the position in buf.
+	// If offset > 0, buf[offset:] has bytes which are read but haven't used.
+	offset int
+	// nextMsgLen is the length of the next log message.
+	// If nextMsgLen = 0, a new value must be read from rdr.
+	nextMsgLen int
 }
 
-func (d *decoder) Decode() (*logger.Message, error) {
-	if d.proto == nil {
-		d.proto = &logdriver.LogEntry{}
-	} else {
-		resetProto(d.proto)
-	}
-	if d.buf == nil {
-		d.buf = make([]byte, initialBufSize)
-	}
-	var (
-		read int
-		err  error
-	)
-
+func (d *decoder) readRecord(size int) error {
+	var err error
 	for i := 0; i < maxDecodeRetry; i++ {
 		var n int
-		n, err = io.ReadFull(d.rdr, d.buf[read:encodeBinaryLen])
+		n, err = io.ReadFull(d.rdr, d.buf[d.offset:size])
+		d.offset += n
 		if err != nil {
 			if err != io.ErrUnexpectedEOF {
-				return nil, errors.Wrap(err, "error reading log message length")
+				return err
 			}
-			read += n
 			continue
 		}
-		read += n
 		break
 	}
 	if err != nil {
-		return nil, errors.Wrapf(err, "could not read log message length: read: %d, expected: %d", read, encodeBinaryLen)
+		return err
 	}
+	d.offset = 0
+	return nil
+}
 
-	msgLen := int(binary.BigEndian.Uint32(d.buf[:read]))
-
-	if len(d.buf) < msgLen+encodeBinaryLen {
-		d.buf = make([]byte, msgLen+encodeBinaryLen)
+func (d *decoder) Decode() (*logger.Message, error) {
+	if d.proto == nil {
+		d.proto = &logdriver.LogEntry{}
 	} else {
-		if msgLen <= initialBufSize {
-			d.buf = d.buf[:initialBufSize]
-		} else {
-			d.buf = d.buf[:msgLen+encodeBinaryLen]
-		}
+		resetProto(d.proto)
+	}
+	if d.buf == nil {
+		d.buf = make([]byte, initialBufSize)
 	}
 
-	return decodeLogEntry(d.rdr, d.proto, d.buf, msgLen)
+	if d.nextMsgLen == 0 {
+		len, err := d.decodeSizeHeader()
+		if err != nil {
+			return nil, err
+		}
```

Comment:
I wonder if the value should be sanity checked here to avoid excessive allocation. `len` is read from a 32bits value.

Reply:
decodeSizeHeader is internally calling binary.BigEndian.Uint32(). So it wouldn't be larger than Unit32.

Reply:
Yep. As you mentioned, it is ultimately depends on the host's memory situation.

Reply:
I think it would be better to check size and error than alloc something that is clearly wrong.
There is a maximum for each message (raw log message max is 1MB, IIRC, not sure what that would be in proto+metadata). We could check something like 512MB, which is way to much to bother allocating.

Reply:
Oh, I didn't know the 1MB cap. Is it defined by the files under docker/daemon/logger?

Reply:
Oh it's even smaller, https://github.com/moby/moby/blob/d456264580cf21de8f22d678afd0cf5decfea36a/daemon/logger/copier.go#L21

Log entries are broken up at newline or 16K.

Reply:
But a line could be longer than 16K. Isn't it?

Reply:
No, 16K is the max and anything longer is broken up into multiple entries.

Reply:
The actual proto we record may be larger than 16K, of course, but that has a soft max just by virtue of the raw log message being capped.